### DISCUSSION TITLE
Why allocateHandle returns offset shifted by MIN_ALIGNMENT_SHIFT?

### DISCUSSION THREAD

**Discussion by michaeltheprogrammer1** [google/filament#3550] [General]
I am investigating backend API and my mind has just got stuck at memory management for graphics handles. In general, I understand how arena and allocator, and free list, etc. work. I don't understand why `HandleBase::HandleId OpenGLDriver::allocateHandle(size_t size) ` returns `offset` shifted by `MIN_ALIGNMENT_SHIFT`, though. And why it cannot just return `addr`. 

whole function:
```cpp
UTILS_NOINLINE
HandleBase::HandleId OpenGLDriver::allocateHandle(size_t size) noexcept {
    void* addr = mHandleArena.alloc(size);
    assert(addr);
    char* const base = (char *)mHandleArena.getArea().begin();
    size_t offset = (char*)addr - base;
    return HandleBase::HandleId(offset >> HandleAllocator::MIN_ALIGNMENT_SHIFT);
}
```

**Comment by romainguy** [comment]
@pixelflinger
